zio-interop-cats/shared/src/main/scala/zio/interop/package.scala

```diff
@@ -34,21 +34,35 @@ package object interop {
   @inline def toEffect[F[_], R, A](rio: RIO[R, A])(implicit R: Runtime[R], F: Async[F], trace: Trace): F[A] =
     F.defer {
       val interrupted = new AtomicBoolean(true)
-      F.async[Exit[Throwable, A]] { cb =>
-        Unsafe.unsafe { implicit unsafe =>
-          val fiber          = R.unsafe.fork {
+      F.asyncCheckAttempt[Exit[Throwable, A]] { cb =>
+        F.delay {
+          implicit val unsafe: Unsafe = Unsafe.unsafe
+
+          val out = R.unsafe.runOrFork {
             signalOnNoExternalInterrupt {
               rio
             }(ZIO.succeed(interrupted.set(false)))
           }
-          fiber.unsafe
-            .addObserver(exit => cb(Right(exit)))
-          val cancelerEffect = F.delay {
-            val _ = fiber.interrupt
+          out match {
+            case Left(fiber) =>
+              val completeCb = (exit: Exit[Throwable, A]) => cb(Right(exit))
+              fiber.unsafe.addObserver(completeCb)
+              Left(Some(F.async[Unit] { cb =>
```

Comment:
I don't think it does. The finalizer is wrapped in `delay` so it's only invoked when cancellation is invoked. Is that explaining what you asked or did I miss something?

Reply:
The async part is so that we don't synchronously block till interruption is completed.

Let's forget about interop for a minute. If we were working purely with ZIO or CIO, wouldn't you expect that all finalisers have finished executing when you issue an interrupt signal and await for the fiber to complete? This is what the async part does in a nutshell; it makes sure that everything in the interrupted effect has finalised by the time we resume execution.